### PR TITLE
fix DistributedDataParallel for VariableImpl to TensorImpl redesign

### PR DESCRIPTION
**Pull request by kmatzen**
Summary: https://github.com/pytorch/pytorch/pull/13827 introduced a change where underlying storage could no longer be changed after calling tensor.data.  There are two places in the DistributedDataParallel class where this requirement is violated.  For now, fix up the DistributedDataParallel class by replacing the inplace storage update with a copy and removing one optimization meant to reduce the memory footprint on replicas.

Differential Revision: D13571669

### CODE REVIEW

**Comment by soumith** [comment]
cc: @teng-li

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -391,7 +391,7 @@ def _sync_params(self):
                                          self.broadcast_bucket_size)
             for tensors, module_params_data in zip(result[1:], self.modules_params_data[1:]):
                 for tensor, param_data in zip(tensors, module_params_data):
-                    param_data.set_(tensor)
```
Comment (teng-li) [inline torch/nn/parallel/distributed.py]: @kmatzen Can you elaborate on why storage could no longer be changed here?  I don't see tensor.data is called on this one

Comment (teng-li) [inline torch/nn/parallel/distributed.py]: Never mind, actually this is a tensor.data.  I wonder how much performance degradation this will cause, since this is on the DDP critical path.

**Comment by teng-li** [comment]
I wonder if you can add a simple test for this?  Like a model wrapped up in DDP (that uses two or more GPUs) that only calls the forward, and verifies that the two or more GPUs have the same parameters.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (kmatzen) [reply torch/nn/parallel/distributed.py]: Is there another way with the redesign to alias tensor storage?  I didn't really understand how to do that anymore.

**Comment by kmatzen** [comment]
Looks like two tests failed due to a failure to download resnet18-5c106cde.pth. Is there a way to trigger a re-run of just these two test configurations?

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (yf225) [reply torch/nn/parallel/distributed.py]: It would be great to measure how much performance degradation this incurs, and if the overhead is significant we can consider adding a `allow_tensor_metadata_change` flag to `.detach()`.

**Comment by pietern** [comment]
@kmatzen Is this still relevant?
